### ISSUE TITLE
No lock files for `bzlmod` modules

### ISSUE THREAD

**Issue by shs96c** [bazelbuild/bazel#14554]
### Description of the problem / feature request:

Although the algorithm used by `bzlmod` should lead to stable results, having to recalculate the dependency tree with each build is wasteful and may prevent reproducible builds, particularly of historic commits. The `WORKSPACE` file locked everything down, so there was no need to worry about this, but it would be useful if `bzlmod` had a lock file it could (optionally) use.

### Feature requests: what underlying problem are you trying to solve with this feature?

I want reproducible builds, even if someone has yanked a module since the last build.

### What's the output of `bazel info release`?

`release 5.0.0rc3`

**Comment by Wyverald** [comment]
This is already on the roadmap. Bzlmod will support a lockfile prior to official launch. (Unclear whether we should leave this issue open -- @meteorcloudy wdyt?)

**Comment by meteorcloudy** [comment]
I think we can leave this issue open to track the progress of lock file implementation.

**Comment by brentleyjones** [comment]
> Bzlmod will support a lockfile prior to official launch.

Is this still on the roadmap?

**Comment by meteorcloudy** [comment]
Definitely, @SalmaSamy is working on this

**Comment by fmeum** [comment]
@SalmaSamy Are you planning to have this go through the regular Bazel proposal process?

**Comment by SalmaSamy** [comment]
@fmeum Yes, we will share the proposal draft soon.

**Comment by meteorcloudy** [comment]
The Bzlmod lockfile proposal is available at https://docs.google.com/document/d/1HPeH_L-lRK54g8A27gv0q7cbk18nwJ-jOq_14XEiZdc/edit#heading=h.5mcn15i0e1ch

**Comment by keith** [comment]
Looks like this just landed https://github.com/bazelbuild/bazel/commit/3b9ec35f5a18afe256f679df1e91796dae720fc3

Minor nit, is there a way the contents could be formatted? Right now it's hard to diff since it's 1 giant line

**Comment by keith** [comment]
Also when building the second time after having created the file I get this exception:

```
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'com.google.devtools.build.lib.bazel.bzlmod.BazelLockFileValue$$Lambda$239/0x000000080028b578@51b4971e' (requested by nodes 'com.google.devtools.build.lib.bazel.bzlmod.BazelDepGraphValue$$Lambda$425/0x00000008005ce2a8@7763b24e')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:592)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:375)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: net.starlark.java.eval.Starlark$InvalidStarlarkValueException: invalid Starlark value: class java.util.ArrayList
```

**Comment by SalmaSamy** [comment]
Hi @keith, A change is made to format the file.
Also, Can you provide a reproducible scenario for this error please?

**Comment by meteorcloudy** [comment]
> *.bazel files are usually written in some dialect of Starlark, but MODULE.lock.bazel is JSON. The proposal included suggestions such as bazel-lock.json, which I think would be more descriptive.

@fmeum @Wyverald @SalmaSamy 
Do you think `MODULE.bazel.lock`would be better? It'll make sure the lock file appears next to the MODULE.bazel, but doesn't make strong implication on its format.

**Comment by Wyverald** [comment]
I'm ambivalent between MODULE.bazel.lock and MODULE.lock.bazel. Either is okay IMO :)

**Comment by meteorcloudy** [comment]
I'm slightly in favor of `MODULE.bazel.lock`, because in reality people easily associate file extension with file format, although technically in our case it's the full file names associated with the format (`WORKSPACE.bazel`, `BUILD.bazel`, `MODULE.bazel` => Starlark).

**Comment by fmeum** [comment]
I do like `MODULE.bazel.lock` more than `MODULE.lock.bazel` for the reasons @meteorcloudy gave. 

I'm not completely sure whether I like it more than a name that doesn't involve `MODULE`: `MODULE.bazel.lock` may give off the impression that it "locks" the current module rather than all other modules (and, in the future, possibly also all external repositories?). For go, `go.sum` probably shouldn't be called `go.mod.lock` for similar reasons.

**Comment by SalmaSamy** [comment]
@fmeum Regarding the other issues:
1. I will look into that.
2. This is very temporary, I am working on the flag mentioned int the document which won't be part of Starlark semantics.
4. An update is pushed for that.

Thanks!

**Comment by fmeum** [comment]
> @fmeum Can you explain more what you mean by "it 'locks' the current module rather than all other modules"?
> We are suppose to eventually lock the current module with its dependencies, overrides, extensions and so on.

The `MODULE.bazel` file, assuming a read-only registry, fully determines all the module dependencies and module extension tags of all transitive Bzlmod dependencies. However, it does not necessarily fully determine the attributes of the repository rules instantiated by module extension. For example, the repository rules instantiated by a `go_sdk` extension could depend on the result of an API call that determines the latest Go SDK version.

In this sense the Bzlmod lock file will in the future pin *more* than just the resolved contents of all `MODULE.bazel` files, it will also pin the results of arbitrary non-hermetic operations performed by module extensions. I am just wondering whether `MODULE.bazel.lock` could give of the impression that it would only do the former. But I still like the name and this may just be a theoretical concern, so feel free to go ahead with it.

**Comment by meteorcloudy** [comment]
Yeah, I guess no matter what name we choose, eventually experienced users will understand what it is for. The difference is basically which one is more friendly to beginner users, and for that I still prefer `MODULE.bazel.lock` a bit more that `bazel-lock.json`.

**Comment by SalmaSamy** [comment]
The name is updated to `MODULE.bazel.lock`.

**Comment by SalmaSamy** [comment]
The flag has been updated to match the document.
Now it is "--lockfile_mode={mode}" where mode can be off, update and error.

@fmeum Can you please provide a reproducible scenario for point #1, and how do you know that resolution was rerun?

**Comment by fmeum** [comment]
@SalmaSamy I cannot reproduce anymore. Looks pretty good now!

**Comment by SalmaSamy** [comment]
@fmeum I see.
That is because we were storing the module registry in the lockfile, so it was working but still reaching to registry to get the repospecs. Now after adding repospec to the module, we don't store/use the registry that anymore.

**Comment by Wyverald** [comment]
This has mostly landed on master and will be in 6.3.0. Closing.
